### PR TITLE
Make pandas required dependency

### LINKED ISSUE #12261: Add pandas as hard dependency
This was agreed in https://github.com/bokeh/bokeh/pull/12020#discussion_r913256527

Do we want it in the impending 3.0 release or not?

### PR DESCRIPTION
**Pull request by bryevdv**
cc @bokeh/core I went ahead and made this PR to be able to see what making pandas non-optional looks like in concrete terms. To me this would be a no-brainer:

* wheels and conda pacakges are readily available nowaday
* also [available on pyodide](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)
* gets rid of the gorpy conditionals throughout
* removes hacky testfixture needed in lots of places, one less hurdle for tests
* no surprises for `hexbin` or sampledata users when pandas is not installed
* some examples will be able to be streamlined and improved
* probably useful to rely on pandas to simplify other features / codepaths

- [x] issues: fixes #12261
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

@mattpap eslint is suddenly failing for me locally, did something change? I have npm  re-installed to no avail.

### CODE REVIEW

File: conda/environment-release-build.yml [hunk 1]
```diff
@@ -4,18 +4,6 @@ channels:
...
-  - jinja2 >=2.7
-  - numpy >=1.11.3
-  - packaging >=16.8
-  - pillow >=4.0
-  - python-dateutil >=2.1
-  - pyyaml >=3.10
-  - tornado >=5
-  - typing_extensions >=3.10.0
-  - xyzservices>=2021.09.1
-
```
Comment (bryevdv) [inline conda/environment-release-build.yml:18]: These should no longer be needed given out isolate setuptools builds, so should be removed as redundant

File: pyproject.toml [hunk 2]
```diff
@@ -13,6 +13,7 @@ dependencies = [
     "contourpy >=1",
     "numpy >=1.11.3",
     "packaging >=16.8",
+    "pandas >=1.2",
```
Comment (bryevdv) [inline pyproject.toml:16]: Pulled this version out of a hat. Open to other suggestions, should we continue.

**Comment by bryevdv** [comment]
I guess the way we are installing the conda package without `--use-local` skips bringing in runtime deps. I added to tests deps explicitly for now, but I'd propose either fixing the conda install method, or installing the wheel artifact instead, and removing the runtime deps from the conda envs to avoid duplication.

File: src/bokeh/core/serialization.py [hunk 3]
```diff
@@ -441,12 +437,11 @@ def _encode_other(self, obj: Any) -> AnyRep:
...
 
-        if _HAS_PANDAS:
-            module = type(obj).__module__
-            if module is not None and module.startswith("pandas"):
-                pd = import_optional("pandas")
-                if pd and isinstance(obj, (pd.Series, pd.Index)):
-                    return self._encode_ndarray(transform_series(obj))
+        import pandas as pd
+        module = type(obj).__module__
+        if module is not None and module.startswith("pandas"):
```
Comment (mattpap) [inline src/bokeh/core/serialization.py:442]: This was guarding the optional import. Either remove it or move the import under this condition. Perhaps the later, unless we want to import pandas on serialization when no pandas objects are involved.

**Comment by mattpap** [comment]
> eslint is suddenly failing for me locally, did something change? 

Failing how? If it's in vscode, just restart the server (F1, "restart ESlint server").

File: src/bokeh/core/serialization.py (hunk 3 above)
Comment (bryevdv) [reply src/bokeh/core/serialization.py:442]: I suppose unlike other places where deferred imports are a result of direct usage involving pandas, serialization will be used regardless. I'll put back the guard but with a comment explaining why it is there.

**Comment by bryevdv** [comment]
@mattpap no I mean eslint is reporting errors when it runs
<details>

```
_________________________________ test_eslint __________________________________

    def test_eslint() -> None:
        ''' Assures that the BokehJS codebase passes configured eslint checks

        '''
        chdir(TOP_PATH/"bokehjs")
        proc = run(["node", "make", "lint"], capture_output=True)
>       assert proc.returncode == 0, f"eslint issues:\n{proc.stdout.decode('utf-8')}"
E       AssertionError: eslint issues:
E         [08:04:51] Starting 'eslint:make'...
E         [08:04:52] Finished 'eslint:make' after 887 ms
E         [08:04:52] Starting 'eslint:lib'...
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/src/lib/core/visuals/hatch.ts
E         [08:04:52]   116:14  warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   116:85  warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   226:51  warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   315:49  warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts
E         [08:04:52]   235:11  warning  Unnecessary conditional, both sides of the expression are literal values  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/src/lib/models/glyphs/image.ts
E         [08:04:52]   22:9  warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/src/lib/models/glyphs/image_base.ts
E         [08:04:52]   135:13  warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   200:9   warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   203:9   warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   206:9   warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]   212:9   warning  Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition
E         [08:04:52]
E         [08:04:52] ✖ 11 problems (0 errors, 11 warnings)
E         [08:04:52] Finished 'eslint:lib' after 70 ms
E         [08:04:52] Starting 'eslint:compiler'...
E         [08:04:52] Finished 'eslint:compiler' after 30 ms
E         [08:04:52] Starting 'eslint:server'...
E         [08:04:52] Finished 'eslint:server' after 27 ms
E         [08:04:52] Starting 'eslint:test:unit'...
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/test/unit/models/glyphs/image_url.ts
E         [08:04:52]   45:14  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/test/unit/models/graphs/graph_hit_test_policy.ts
E         [08:04:52]   100:16  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]   173:16  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/test/unit/models/tools/inspectors/hover_tool.ts
E         [08:04:52]   254:12  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/test/unit/models/widgets/checkbox_button_group.ts
E         [08:04:52]   36:7  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]
E         [08:04:52] ✖ 5 problems (5 errors, 0 warnings)
E         [08:04:52]   5 errors and 0 warnings potentially fixable with the `--fix` option.
E         [08:04:52] Finished 'eslint:test:unit' after 40 ms
E         [08:04:52] failed: lint failed with 5 errors
E         [08:04:52] failed: lint failed with 5 errors
E         [08:04:52] Starting 'eslint:test:defaults'...
E         [08:04:52] Finished 'eslint:test:defaults' after 22 ms
E         [08:04:52] Starting 'eslint:test:integration'...
E         [08:04:52] Finished 'eslint:test:integration' after 24 ms
E         [08:04:52] Starting 'eslint:test:codebase'...
E         [08:04:52] Finished 'eslint:test:codebase' after 24 ms
E         [08:04:52] Starting 'eslint:test:devtools'...
E         [08:04:52] Finished 'eslint:test:devtools' after 22 ms
E         [08:04:52] failed: task 'top-level' failed because 'eslint:test:unit' failed
E         [08:04:52] failed: task 'top-level' failed because 'eslint:test:unit' failed
E         [08:04:52] Starting 'eslint:examples'...
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/examples/hover/hover.ts
E         [08:04:52]   49:3  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]
E         [08:04:52] /Users/bryan/work/bokeh/bokehjs/examples/tap/tap.ts
E         [08:04:52]   49:3  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]   50:3  error  This assertion is unnecessary since it does not change the type of the expression  @typescript-eslint/no-unnecessary-type-assertion
E         [08:04:52]
E         [08:04:52] ✖ 3 problems (3 errors, 0 warnings)
E         [08:04:52]   3 errors and 0 warnings potentially fixable with the `--fix` option.
E         [08:04:52] Finished 'eslint:examples' after 27 ms
E         [08:04:52] failed: lint failed with 3 errors
E         [08:04:52] failed: lint failed with 3 errors
E         [08:04:52] failed: task 'top-level' failed because 'eslint:test' and 'eslint:examples' failed
E         [08:04:52] failed: task 'top-level' failed because 'eslint:test' and 'eslint:examples' failed
E         [08:04:52] failed: task 'top-level' failed because 'eslint' failed
E         [08:04:52] failed: task 'top-level' failed because 'eslint' failed
E         [08:04:52] failed: task 'top-level' failed because 'lint' failed
E         [08:04:52] failed: task 'top-level' failed because 'lint' failed
E
E       assert 1 == 0
E        +  where 1 = CompletedProcess(args=['node', 'make', 'lint'], returncode=1, stdout=b"[08:04:51] Starting 'eslint:make'...\n[08:04:52...p-level' failed because 'lint' failed\n[08:04:52] failed: task 'top-level' failed because 'lint' failed\n", stderr=b'').returncode

/Users/bryan/work/bokeh/tests/codebase/test_eslint.py:37: AssertionError
```

</details>

**Comment by bryevdv** [comment]
I'm not sure what's up with the mypy errors, some of them just appear to be wrong or bad type stubs, etc. No sure why they are only just appearing.

**Comment by mattpap** [comment]
Try removing `bokehjs/.eslintcache`, that usually helps when eslint goes crazy.

**Comment by mattpap** [comment]
> I'm not sure what's up with the mypy errors, some of them just appear to be wrong or bad type stubs, etc. No sure why they are only just appearing.

Mostly bad typings and either bug in mypy or mypy being overly cautious. They appear now, because previously many code paths related to pandas were dynamically typed, because there was no way to type `import_optional()` (everything after that was `Any`).

**Comment by bryevdv** [comment]
cc @bokeh/core interested in further thoughts:

* is 1.2 a good pandas min ver? 
* @philippjfr any concerns with adding pandas wrt  panel/hv?
* @tcmetzger any other places that mention optional pandas? I removed the one I know about

**Comment by bryevdv** [comment]
@bokeh/core I intend to merge this after tomorrow's meeting, unless there is comment before then.

**Comment by jrbourbeau** [comment]
I don't think should be problematic on the `dask` side of things 👍

**Pull request merged**